### ISSUE TITLE
Flat State Cache Sizing

### ISSUE THREAD

**Issue by jakmeier** [near/nearcore#8684]
Issue to track performance analysis done to fine-tune sizing of the different caches involved with Flat Storage.

We want balance RAM requiremens and performance gained from caches.

Caches we could use are:
- RocksDB Column Caches (`DBCol::State`, `DBCol::FlatState`, `DBCol::BlockInfo`, `DBCol::BlockHeight`)
- RocksDB Row Cache (`DBCol::FlatState`)
- Flat State Key-ValueRef cache, with the option to limit it by key length
- Trie Shard Cache (for Values shared between flat and trie storage, as well as trie-only ValueRefs)
- `FlatStateDelta` as in-memory view to deltas that are also stored on disk

The baseline we have been using so far is:
- RocksDB Column Caches (+512MB)
  - `DBCol::State` | `DBCol::FlatState` => 512MB
  - `DBCol::BlockInfo` | `DBCol::BlockHeight` => 64MB
- RocksDB Row Cache disabled (+0MB)
- Flat State Key-ValueRef cache (+800MB)
  - capacity: 4x100_000
  - max key length = 2000B
- Trie Shard Cache (+0MB)
  - 50MB default per shard
  - shard 3 has 3GB
- `FlatStateDelta` is technically unlimited but we assume it will stay below 600MB total (+600MB)

## Summary of results and progress

- [x] Trie cache: We can reduce the shard 3 size to 1GB
- [x] RocksDB `DBCol::FlatState Column Cache: is 512MB ok? => Yes, although we could also consider lowering it.
- [x] RocksDB Row Cache: enabling it makes no sense, all experiments have shown significantly _worse_ performance
- [x] Flat state cache: Analysis done, decided to to remove it.
- [x] do we still need to increase RocksDB Column Caches for `DBCol::BlockInfo`, `DBCol::BlockHeight`? => No, submitted #8726

**Comment by jakmeier** [comment]
# Tuning the FlatState ValueRef cache and the shard cache
**TLDR**
- Can we reduce the 3GB of the s3 shard cache?
  - Yes, we can go to 1GB. But validators and indexers may choose to keep 3GB.
- Can we reduce the 4x100_000 capacity of FS cache?
  - Yes, 100k is way oversized. We can get about 40% of the possible hits with a capacity of 10. (Without a k!)
  - In fact, we can probably remove the cache completely.
- Does a smaller key limit for the FS cache make a difference?
  -  A limit of 200 instead of 2000 makes no difference (100% of the possible hits)
  - Also, a limit of 50 still gives around 60% of the possible hits

## Methodology
To analyze 3 parameters (Trie chace, FS cache, FS cache key limit), I run multiple mainnet nodes. They have to run simultaneously because the workload at different times is not directly comparable.

I chose to use archival nodes, because they are most sensitive to storage latency. Plus, if we can show that a setting works for archival nodes, we can be confident that both validators and archival nodes can keep up. It might give more conservative estimates than strictly necessary, since we mostly care about validators and not archival nodes. But I think that's ok, better to be safe than sorry.

One way to analyze all parameters would be change each parameter individually and compare it to a baseline. But we want to change multiple of them and we also expect a strong interaction between the second and third parameter. So to get meaningful results, I want to measure the effect of each parameter in general, not only against an arbitrary baseline.

The standard soultion to that would be a [Full Factorial Experiement](https://en.wikipedia.org/wiki/Factorial_experiment). However, testing 3 parameters at N levels gives N^3 combinations. And because I want to run them simultaneously, it would require that many machines to cover all configurations. And archival nodes are not cheap.

To reduce the number of machines I need, I decided to go with only 2 parameter levels (N=2). Specifically, take each value that we have been using so far as the high level, and compare it to a 10x reduce value as the lower level. That should be enough for an initial assessment of which parameters are important. Fine tuning can follow once we know which parameters matter.

 But 2^3 = 8  is still a non-trivial number of archival nodes to spin up. So I went with a 2^(3-1) [Fractional Factorial Design](https://en.wikipedia.org/wiki/Fractional_factorial_design) to cut it in half. This way I only need 4 machines and can still evaluate the importance of all three parameters.

The last choice I made was to look primarily at block processing time. There are more detailed metrics to look at, such as time per DB column, or number of requests per column. But these are harder to compare. It's better to only use them to dig deeper but not for the overall performance comparison.

## First Results

The fractional factorial analysis is available here: https://docs.google.com/spreadsheets/d/1rMIZTJyg81yUJbWKLFIAiUd0rQ1mEZfP1c36RgncRVM/edit?usp=sharing

The table below summarizes the effect each parameter had in the experiment, therefore giving an idea of how important each parameter is for overall performance.

Allocation of Variance in block processing time  | Steady state (>48h)
-- | --
L2 trie node cache (shard3) | 90.19%
FS cache capacity | 0.72%
FS cache max key size  | 4.15%
Experimental model error  | 4.94%

From this, I conclude that the L2 trie node cache is very important for block processing time.
The FS cache has a very minor effect and especially the difference between 100k or 10k slots capacity is virtually zero.
The key limit did make a small difference but it's hard to explain why exactly.

An equally important result is that even the slow setups were able to keep up with the network. Hence, all optimizations are beyond the strict minimum.

## Follow-p experiments
So I played around with other FS cache sizes and key limits. Hit rates can be as high as 80% but only when there are a small number of requests. It is better to look at absolute hit and miss numbers.

  | flat state cache | flat state max key size | idle [miss/min] | idle[hit/min] | Sweatcoin bursts [miss/min] | Sweatcoin bursts [hit/min]
-- | -- | -- | -- | -- | -- | --
no FS cache | 0 | 0 | 60 | 0 | 807 | 0
large FS cache | 100000 | 2000 | 13 | 47 | 742 | 65
hard key limit | 100000 | 50 | 30 | 30 | 758 | 49
tiny cache | 10 | 2000 | 40 | 20 | 792 | 15

The table shows that FS cache sizing has an impact on hits and misses, and so does the key limit. But we can reduce the capacity from 100000 to just 10 (reduction of 10k times) and still get around 40% of the hits.

However, ultimately, none of this really matters because we optimize 1 or 2 DB requests per block. And each FlatState DB request only takes around 20us. We should consider removing the FS cache entirely.

**Comment by jakmeier** [comment]
Submitted #8726 to address the latest findings. I believe this (and the removal of the FS cache) are the only changes I want to make to cache sizes.

**Comment by jakmeier** [comment]
With #8723 and #8727 merged, I will close this issue as done.

**Comment by Longarithm** [comment]
Somehow we missed change for RPC/archival nodes on config side. Pinging SREs about this.

**Comment by jakmeier** [comment]
I thought we want to increase by default and then tell validators that they can lower it to reduce RAM usage. See this Zulip topic: https://near.zulipchat.com/#narrow/stream/297873-pagoda.2Fnode/topic/min.20vs.20recommended.20node.20setup
